### PR TITLE
[inductor] autotune benchmark support for cpu

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124021
* #125736
* __->__ #125159

This PR adds the autotune Infrastructure for CPU. It generalizes and extends `BenchmarkRequest` with CPU support and C++ module loader. A `do_bench_cpu` util function is added for benchmarking functions on CPU with warmups and returns the median number from multiple trials.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: third_party/ideep
Comment (jansel) [inline third_party/ideep:1]: Do submodule update in a different PR.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3627,7 +3628,10 @@ def __init__(self, name, input_nodes, layout):
 
     def benchmark(self, *args, out) -> float:
         algo = self.to_callable()
-        return do_bench(lambda: algo(*args, out=out))
+        if is_cpu_device(args):
+            return do_bench_cpu(lambda: algo(*args, out=out))
+        else:
+            return do_bench(lambda: algo(*args, out=out))
```
Comment (shunting314) [inline torch/_inductor/ir.py:3634]: Can we have this device dispatch logic inside do_bench?

Basically, rename current do_bench to do_bench_gpu. And create a new do_bench function that dispatches to do_bench_cpu/do_bench_gpu based on devices.

File: torch/_inductor/autotune_process.py [hunk 2]
```diff
@@ -763,6 +791,71 @@ def __str__(self) -> str:
...
+    def do_bench(
+        self,
+        fn,
+        *input_tensors: torch.Tensor,
+        output_tensor: Optional[torch.Tensor] = None,
+    ) -> float:
+        return do_bench_cpu(fn)
+
+
+class CppBenchmarkRequest(CPUDeviceBenchmarkRequest):
```
Comment (shunting314) [inline torch/_inductor/autotune_process.py:804]: hmm, where is this being used?

File: torch/_inductor/ir.py (hunk 1 above)
Comment (jgong5) [reply torch/_inductor/ir.py:3634]: Thanks. I did the refactoring in a separate PR: https://github.com/pytorch/pytorch/pull/125736

File: third_party/ideep
Comment (jgong5) [reply third_party/ideep:1]: Thanks. A wrong "git add". Fixed.

**Review by jansel — APPROVED** [review]

**Comment by jerryzh168** [comment]
we do have api depend on `do_bench` btw: https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L200 is this not a public API?

**Comment by jgong5** [comment]
> we do have api depend on `do_bench` btw: https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L200 is this not a public API?

I thought it was only used by inductor internal. You may have to rename it to `do_bench_gpu` now since the usage is cuda specific?
